File: buzzing/lib/page/meeting/signaling/signaling.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:buzzing/utils/logger_util.dart';
import 'package:flutter/material.dart';
import 'package:flutter_macos_permissions/flutter_macos_permissions.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart' as rtc;

import 'device_info.dart';
import 'session.dart';
import 'simple_ws.dart';

enum SignalingState { ConnectionOpen, ConnectionClosed, ConnectionError }

enum CallState {
  CallStateNew,
  CallStateRinging,
  CallStateInvite,
  CallStateConnected,
  CallStateBye,
}

enum VideoSource { Camera, Screen }

class SignalingConfig {
  final String host;
  final int port;
  final String token;
  Map<String, dynamic> iceServers;

  SignalingConfig({
    this.host = 'www.buzzing-im.com',
    this.port = 5150,
    required this.token,
    Map<String, dynamic>? iceServers,
  }) : iceServers = iceServers ?? {
          'iceServers': [
            {'url': 'stun:stun.l.google.com:19302'},
          ],
        };
}

class Signaling {
  Signaling(this.context, {required String uid, required String token})
      : uid = uid,
        config = SignalingConfig(token: token);

  final JsonEncoder encoder = JsonEncoder();
  final JsonDecoder decoder = JsonDecoder();
  late String uid;
  SimpleWebSocket? socket;
  BuildContext? context;

  late SignalingConfig config;
  var turnCredential;
  final Map<String, Session> sessions = {};
  MediaStream? localStream;
  final List<MediaStream> remoteStreams = <MediaStream>[];
  final List<RTCRtpSender> senders = <RTCRtpSender>[];
  VideoSource videoSource = VideoSource.Camera;

  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;
  Timer? _keepaliveTimer;
  Timer? _keepaliveTimeout;
  bool _intentionalClose = false;
  bool _connected = false;

  Function(SignalingState state)? onSignalingStateChange;
  Function(Session session, CallState state)? onCallStateChange;
  Function(MediaStream stream)? onLocalStream;
  Function(Session session, MediaStream stream)? onAddRemoteStream;
  Function(Session session, MediaStream stream)? onRemoveRemoteStream;
  Function(dynamic event)? onPeerUpdate;
  Function(Session session, RTCDataChannel dc, RTCDataChannelMessage data)?
      onDataChannelMessage;
  Function(Session session, RTCDataChannel dc)? onDataChannel;
  Function()? onReconnect;

  String get sdpSemantics => 'unified-plan';

  static const Map<String, dynamic> peerConfig = {
    'mandatory': {},
    'optional': [
      {'DtlsSrtpKeyAgreement': true},
    ],
  };

  static const Map<String, dynamic> dcConstraints = {
    'mandatory': {'OfferToReceiveAudio': false, 'OfferToReceiveVideo': false},
    'optional': [],
  };

  close() async {
    L.d('RTC close');
    _intentionalClose = true;
    _stopReconnect();
    _stopKeepalive();
    await cleanSessions();
    socket?.close();
    socket = null;
  }

  void switchCamera() {
    L.d('RTC switch camera');
    if (localStream != null) {
      if (videoSource != VideoSource.Camera) {
        senders.forEach((sender) {});
        videoSource = VideoSource.Camera;
        onLocalStream?.call(localStream!);
      } else {
        Helper.switchCamera(localStream!.getVideoTracks()[0]);
      }
    }
  }

  Future<void> startScreenSharing() async {
    if (videoSource == VideoSource.Screen) return;
    L.d('RTC start screen sharing');
    try {
      final stream = await rtc.navigator.mediaDevices.getDisplayMedia({
        'video': {
          'mandatory': {'minWidth': '1280', 'minHeight': '720'},
        },
        'audio': false,
      });
      if (localStream != null && stream.getVideoTracks().isNotEmpty) {
        for (final sender in senders) {
          if (sender.track?.kind == 'video') {
            await sender.replaceTrack(stream.getVideoTracks()[0]);
          }
        }
        stream.getVideoTracks()[0].onEnded = () {
          L.d('screen sharing stopped by system');
          stopScreenSharing();
        };
        videoSource = VideoSource.Screen;
        onLocalStream?.call(stream);
      }
    } catch (e) {
      L.d('start screen sharing error: $e');
    }
  }

  Future<void> stopScreenSharing() async {
    if (videoSource != VideoSource.Screen) return;
    L.d('RTC stop screen sharing');
    if (localStream != null) {
      final cameraStream = await _createCameraStream();
      for (final sender in senders) {
        if (sender.track?.kind == 'video' && cameraStream.getVideoTracks().isNotEmpty) {
          await sender.replaceTrack(cameraStream.getVideoTracks()[0]);
        }
      }
      videoSource = VideoSource.Camera;
      onLocalStream?.call(cameraStream);
    }
  }

  void muteMic() {
    L.d('RTC mute mic');
    if (localStream != null) {
      final enabled = localStream!.getAudioTracks()[0].enabled;
      localStream!.getAudioTracks()[0].enabled = !enabled;
    }
  }

  void invite(String peerId, String media, bool useScreen) async {
    L.d('RTC invite, $peerId, $media, $useScreen');
    var sessionId = '$uid-$peerId';
    Session session = await createSession(
      null,
      peerId: peerId,
      sessionId: sessionId,
      media: media,
      screenSharing: useScreen,
    );
    sessions[sessionId] = session;
    if (media == 'data') {
      createDataChannel(session);
    }
    createOffer(session, media);
    onCallStateChange?.call(session, CallState.CallStateNew);
    onCallStateChange?.call(session, CallState.CallStateInvite);
  }

  void bye(String sessionId) {
    L.d('RTC send bye');
    send('bye', {'session_id': sessionId, 'from': uid});
    var sess = sessions[sessionId];
    if (sess != null) {
      closeSession(sess);
    }
  }

  void accept(String sessionId, String media) {
    L.d('RTC accept session, $sessionId, $media');
    var session = sessions[sessionId];
    if (session == null) return;
    createAnswer(session, media);
  }

  void reject(String sessionId) {
    L.d('RTC reject session, $sessionId');
    var session = sessions[sessionId];
    if (session == null) return;
    bye(session.sid);
  }

  void onMessage(message) async {
    Map<String, dynamic> mapData = message;
    L.d('RTC receive message, $mapData');
    var data = mapData['data'];
    switch (mapData['type'] as String) {
      case 'peers':
        final List<dynamic> peers = data;
        if (onPeerUpdate != null) {
          onPeerUpdate?.call({'self': uid, 'peers': peers});
        }
        break;
      case 'offer':
        {
          var peerId = data['from'];
          var description = data['description'];
          var media = data['media'];
          var sessionId = data['session_id'];
          var session = sessions[sessionId];
          var newSession = await createSession(
            session,
            peerId: peerId,
            sessionId: sessionId,
            media: media,
            screenSharing: false,
          );
          sessions[sessionId] = newSession;
          L.d('prepare for new session: $newSession');
          await newSession.pc!.setRemoteDescription(
            RTCSessionDescription(description['sdp'], description['type']),
          );
          if (newSession.remoteCandidates.isNotEmpty) {
            for (var candidate in newSession.remoteCandidates) {
              await newSession.pc?.addCandidate(candidate);
            }
            newSession.remoteCandidates.clear();
          }
          onCallStateChange?.call(newSession, CallState.CallStateNew);
          onCallStateChange?.call(newSession, CallState.CallStateRinging);
        }
        break;
      case 'answer':
        {
          var description = data['description'];
          var sessionId = data['session_id'];
          var session = sessions[sessionId];
          await session?.pc?.setRemoteDescription(
            RTCSessionDescription(description['sdp'], description['type']),
          );
          onCallStateChange?.call(session!, CallState.CallStateConnected);
        }
        break;
      case 'candidate':
        {
          var peerId = data['from'];
          var candidateMap = data['candidate'];
          var sessionId = data['session_id'];
          var session = sessions[sessionId];
          var candidate = RTCIceCandidate(
            candidateMap['candidate'],
            candidateMap['sdpMid'],
            candidateMap['sdpMLineIndex'],
          );
          if (session != null) {
            if (session.pc != null) {
              await session.pc?.addCandidate(candidate);
            } else {
              session.remoteCandidates.add(candidate);
            }
          } else {
            sessions[sessionId] = Session(pid: peerId, sid: sessionId)
              ..remoteCandidates.add(candidate);
          }
        }
        break;
      case 'leave':
        clossSessionByPeerId(data as String);
        break;
      case 'bye':
        {
          var sessionId = data['session_id'];
          L.d('meeting bye: $sessionId');
          var session = sessions.remove(sessionId);
          if (session != null) {
            onCallStateChange?.call(session, CallState.CallStateBye);
            closeSession(session);
          }
        }
        break;
      case 'keepalive':
        L.d('meeting conn keepalive response');
        _keepaliveTimeout?.cancel();
        break;
    }
  }

  Future<void> connect() async {
    _intentionalClose = false;
    _reconnectAttempts = 0;
    _stopReconnect();

    var url = 'https://${config.host}:${config.port}/ws';
    socket?.close();
    socket = SimpleWebSocket(url);
    L.d('connect to url: $url');

    if (turnCredential == null) {
      try {
        turnCredential = await getTurnCredential(config.host, config.port, token: config.token);
        if (turnCredential['urls'] != null) {
          config.iceServers = {
            'iceServers': [
              {
                'urls': turnCredential['urls'][0],
                'username': turnCredential['username'],
                'credential': turnCredential['credential'],
              },
            ],
          };
        }
      } catch (_) {}
    }

    socket?.onOpen = () {
      L.d('meeting connect on open');
      _connected = true;
      _reconnectAttempts = 0;
      onSignalingStateChange?.call(SignalingState.ConnectionOpen);
      send('new', {
        'token': config.token,
        'name': DeviceInfo.label,
        'user_agent': DeviceInfo.userAgent,
      });
      _startKeepalive();
    };
    socket?.onMessage = (message) {
      L.d('meeting received data: $message');
      onMessage(decoder.convert(message));
    };
    socket?.onClose = (int? code, String? reason) {
      L.d('meeting conn closed by server, [$code => $reason]');
      _connected = false;
      _stopKeepalive();
      onSignalingStateChange?.call(SignalingState.ConnectionClosed);
      if (!_intentionalClose) {
        _reconnect();
      }
    };
    await socket?.connect();
  }

  void _reconnect() {
    if (_intentionalClose) return;
    _stopReconnect();
    final delay = _reconnectDelay();
    _reconnectAttempts++;
    L.d('reconnect attempt $_reconnectAttempts in ${delay}ms');
    _reconnectTimer = Timer(Duration(milliseconds: delay), () {
      onSignalingStateChange?.call(SignalingState.ConnectionError);
      connect();
    });
  }

  int _reconnectDelay() {
    const maxDelay = 16000;
    int delay = 1000;
    for (int i = 1; i < _reconnectAttempts; i++) {
      delay *= 2;
    }
    return delay > maxDelay ? maxDelay : delay;
  }

  void _stopReconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  void _startKeepalive() {
    _stopKeepalive();
    _keepaliveTimer = Timer.periodic(const Duration(seconds: 3), (_) {
      if (!_connected) return;
      send('keepalive', {});
      _keepaliveTimeout?.cancel();
      _keepaliveTimeout = Timer(const Duration(seconds: 5), () {
        L.d('keepalive timeout, triggering reconnect');
        _connected = false;
        socket?.close();
      });
    });
  }

  void _stopKeepalive() {
    _keepaliveTimer?.cancel();
    _keepaliveTimer = null;
    _keepaliveTimeout?.cancel();
    _keepaliveTimeout = null;
  }

  void _setupSimulcastEncodings(RTCRtpSender sender) async {
    try {
      final params = sender.parameters;
      params.encodings = [
        RTCRtpEncoding(
          rid: 'q',
          active: true,
          scaleResolutionDownBy: 4.0,
          maxBitrate: 100000,
        ),
        RTCRtpEncoding(
          rid: 'h',
          active: true,
          scaleResolutionDownBy: 2.0,
          maxBitrate: 500000,
        ),
        RTCRtpEncoding(
          rid: 'f',
          active: true,
          scaleResolutionDownBy: 1.0,
          maxBitrate: 1500000,
        ),
      ];
      await sender.setParameters(params);
      L.d('simulcast encodings set up');
    } catch (e) {
      L.d('simulcast setup error (non-fatal): $e');
    }
  }

  void _onIceConnectionState(RTCIceConnectionState state, Session session) async {
    L.d('ICE connection state: $state');
    if (!senders.any((s) => s.track?.kind == 'video')) return;
    try {
      final videoSender = senders.firstWhere((s) => s.track?.kind == 'video');
      final params = videoSender.parameters;
      final encodings = params.encodings;
      if (encodings == null || encodings.isEmpty) return;
      switch (state) {
        case RTCIceConnectionState.RTCIceConnectionStateConnected:
        case RTCIceConnectionState.RTCIceConnectionStateCompleted:
          for (final e in encodings) {
            e.active = true;
          }
          break;
        case RTCIceConnectionState.RTCIceConnectionStateDisconnected:
        case RTCIceConnectionState.RTCIceConnectionStateFailed:
          for (final e in encodings) {
            e.active = false;
          }
          break;
        default:
          break;
      }
      await videoSender.setParameters(params);
    } catch (e) {
      L.d('BWE update error: $e');
    }
  }

  Future<MediaStream> createStream(
    String media,
    bool userScreen, {
    BuildContext? context,
  }) async {
    if (userScreen) {
      final stream = await rtc.navigator.mediaDevices.getDisplayMedia({
        'video': {
          'mandatory': {'minWidth': '1280', 'minHeight': '720'},
        },
        'audio': false,
      });
      onLocalStream?.call(stream);
      return stream;
    }
    return _createCameraStream();
  }

  Future<MediaStream> _createCameraStream() async {
    await FlutterMacosPermissions.requestCamera();
    var devices = await rtc.navigator.mediaDevices.enumerateDevices();
    for (var d in devices) {
      L.d('RTC device: ${d.label}, ${d.deviceId}, ${d.kind}');
    }
    final Map<String, dynamic> mediaContraints = {
      'video': {
        'mandatory': {
          'minWidth': '640',
          'minHeight': '480',
          'minFrameRate': '30',
        },
        'facingMode': 'user',
        'optional': [],
      },
      'audio': {
        'echoCancellation': true,
        'noiseSuppression': true,
        'autoGainControl': true,
      },
    };
    var stream = await rtc.navigator.mediaDevices.getUserMedia(mediaContraints);
    onLocalStream?.call(stream);
    return stream;
  }

  Future<Session> createSession(
    Session? session, {
    required String peerId,
    required String sessionId,
    required String media,
    required bool screenSharing,
  }) async {
    var newSession = session ?? Session(sid: sessionId, pid: peerId);
    if (media != 'data') {
      localStream = await createStream(media, screenSharing, context: context);
    }
    L.d('create session, ice server: ${config.iceServers}');
    RTCPeerConnection pc = await createPeerConnection({
      ...config.iceServers,
      ...{'sdpSemantics': sdpSemantics},
    }, peerConfig);
    if (media != 'data') {
      switch (sdpSemantics) {
        case 'plan-b':
          pc.onAddStream = (MediaStream stream) {
            onAddRemoteStream?.call(newSession, stream);
            remoteStreams.add(stream);
          };
          await pc.addStream(localStream!);
          break;
        case 'unified-plan':
          pc.onTrack = (event) {
            if (event.track.kind == 'video') {
              onAddRemoteStream?.call(newSession, event.streams[0]);
            }
          };
          localStream!.getTracks().forEach((track) async {
            final sender = await pc.addTrack(track, localStream!);
            senders.add(sender);
            if (track.kind == 'video') {
              _setupSimulcastEncodings(sender);
            }
          });
          break;
      }
    }

    pc.onIceCandidate = (candidate) async {
      await Future.delayed(
        const Duration(seconds: 1),
        () => send('candidate', {
          'to': peerId,
          'from': uid,
          'candidate': {
            'sdpMLineIndex': candidate.sdpMLineIndex,
            'sdpMid': candidate.sdpMid,
            'candidate': candidate.candidate,
          },
          'session_id': sessionId,
        }),
      );
    };

    pc.onIceConnectionState = (state) {
      _onIceConnectionState(state, newSession);
    };
    pc.onRemoveStream = (stream) {
      onRemoveRemoteStream?.call(newSession, stream);
      remoteStreams.removeWhere((it) => it.id == stream.id);
    };

    pc.onDataChannel = (channel) {
      addDataChannel(newSession, channel);
    };

    newSession.pc = pc;
    return newSession;
  }

  void addDataChannel(Session session, RTCDataChannel channel) {
    channel.onDataChannelState = (e) {};
    channel.onMessage = (RTCDataChannelMessage data) {
      onDataChannelMessage?.call(session, channel, data);
    };
    session.dc = channel;
    onDataChannel?.call(session, channel);
  }

  Future<void> createDataChannel(
    Session session, {
    label = 'fileTransfer',
  }) async {
    RTCDataChannelInit dataChannelDict = RTCDataChannelInit()
      ..maxRetransmits = 30;
    RTCDataChannel channel = await session.pc!.createDataChannel(
      label,
      dataChannelDict,
    );
    addDataChannel(session, channel);
  }

  Future<void> createOffer(Session session, String media,
      {bool iceRestart = false}) async {
    try {
      final constraints = Map<String, dynamic>.from(
          media == 'data' ? dcConstraints : {});
      if (iceRestart) {
        constraints['iceRestart'] = true;
      }
      RTCSessionDescription s = await session.pc!.createOffer(constraints);
      await session.pc!.setLocalDescription(fixSdp(s));
      send('offer', {
        'to': session.pid,
        'from': uid,
        'description': {'sdp': s.sdp, 'type': s.type},
        'session_id': session.sid,
        'media': media,
      });
    } catch (e) {
      L.d('create offer error: $e');
    }
  }

  RTCSessionDescription fixSdp(RTCSessionDescription s) {
    s.sdp = s.sdp!.replaceAll(
      'profile-level-id=640c1f',
      'profile-level-id=42e032',
    );
    return s;
  }

  Future<void> createAnswer(Session session, String media) async {
    try {
      RTCSessionDescription s = await session.pc!.createAnswer(
        media == 'data' ? dcConstraints : {},
      );
      await session.pc!.setLocalDescription(fixSdp(s));
      send('answer', {
        'to': session.pid,
        'from': uid,
        'description': {'sdp': s.sdp, 'type': s.type},
        'session_id': session.sid,
      });
    } catch (e) {
      L.d('create answer error: $e');
    }
  }

  send(event, data) {
    var request = <String, dynamic>{};
    request['type'] = event;
    request['data'] = data;
    socket?.send(encoder.convert(request));
  }

  Future<void> cleanSessions() async {
    if (localStream != null) {
      for (var element in localStream!.getTracks()) {
        await element.stop();
      }
      await localStream!.dispose();
      localStream = null;
    }
    for (var sess in sessions.values) {
      await sess.pc?.close();
      await sess.dc?.close();
    }
    sessions.clear();
  }

  void clossSessionByPeerId(String peerId) {
    Session? session;
    sessions.removeWhere((String key, Session sess) {
      var ids = key.split('_');
      session = sess;
      return peerId == ids[0] || peerId == ids[1];
    });
    if (session != null) {
      closeSession(session);
      onCallStateChange?.call(session!, CallState.CallStateBye);
    }
  }

  void closeSession(Session? session) async {
    if (session == null) return;
    for (var element in (localStream?.getTracks() ?? [])) {
      await element.stop();
    }
    await localStream?.dispose();
    localStream = null;

    await session.pc?.close();
    await session.dc?.close();
    senders.clear();
    videoSource = VideoSource.Camera;
  }
}

Future<Map> getTurnCredential(String host, int port, {String? token}) async {
  HttpClient client = HttpClient(context: SecurityContext());
  client.badCertificateCallback =
      (X509Certificate cert, String host, int port) {
    L.d('getTurnCredential: Allow self-signed certificate => $host:$port.');
    return true;
  };
  var queryParams = 'token=$token';
  var url = 'https://$host:$port/api/turn?$queryParams';
  var request = await client.getUrl(Uri.parse(url));
  var response = await request.close();
  var responseBody = await response.transform(Utf8Decoder()).join();
  L.d('getTurnCredential:response => $responseBody.');
  return JsonDecoder().convert(responseBody);
}

File: frontend/src/services/meeting/signaling.ts
import { useAuthStore } from '@/stores/auth'

export type SignalingState = 'open' | 'closed' | 'error'
export type CallState = 'new' | 'ringing' | 'invite' | 'connected' | 'bye'

interface Session {
  sid: string
  pid: string
  pc: RTCPeerConnection | null
  media: string
  remoteCandidates: RTCIceCandidateInit[]
}

export class Signaling {
  uid = ''
  private ws: WebSocket | null = null
  private host: string
  private port: number
  private token: string
  private sessions = new Map<string, Session>()
  private reconnectAttempts = 0
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null
  private keepaliveTimer: ReturnType<typeof setInterval> | null = null
  private keepaliveTimeout: ReturnType<typeof setTimeout> | null = null
  private intentionalClose = false
  private connected = false
  private _localStream: MediaStream | null = null
  private _remoteStream: MediaStream | null = null

  onSignalingStateChange: ((s: SignalingState) => void) | null = null
  onCallStateChange: ((sid: string, s: CallState) => void) | null = null
  onLocalStream: ((s: MediaStream) => void) | null = null
  onRemoteStream: ((s: MediaStream) => void) | null = null
  onPeerUpdate: ((data: { self: string; peers: unknown[] }) => void) | null = null
  onReconnect: (() => void) | null = null

  get localStream() { return this._localStream }
  get remoteStream() { return this._remoteStream }

  constructor() {
    const auth = useAuthStore()
    this.host = window.location.hostname
    this.port = 5150
    this.token = auth.token || ''
  }

  private wsUrl() {
    return `wss://${this.host}:${this.port}/ws`
  }

  async getTurnCredential(): Promise<RTCIceServer[] | null> {
    try {
      const res = await fetch(`https://${this.host}:${this.port}/api/turn?token=${this.token}`)
      const data = await res.json()
      if (data.urls) {
        return [{
          urls: data.urls,
          username: data.username,
          credential: data.credential,
        }]
      }
    } catch { /* ignore */ }
    return null
  }

  connect() {
    this.intentionalClose = false
    this.reconnectAttempts = 0
    if (this.reconnectTimer) { clearTimeout(this.reconnectTimer); this.reconnectTimer = null }
    if (this.ws) { this.ws.onclose = null; this.ws.close() }

    this.ws = new WebSocket(this.wsUrl())
    this.ws.onopen = () => {
      this.connected = true
      this.reconnectAttempts = 0
      this.onSignalingStateChange?.('open')
      this.send('new', { token: this.token, name: 'web', user_agent: navigator.userAgent })
      this.startKeepalive()
    }
    this.ws.onmessage = (e) => {
      try {
        const msg = JSON.parse(e.data)
        this.onMessage(msg)
      } catch { /* ignore */ }
    }
    this.ws.onclose = () => {
      this.connected = false
      this.stopKeepalive()
      this.onSignalingStateChange?.('closed')
      if (!this.intentionalClose) this.reconnect()
    }
    this.ws.onerror = () => {
      this.onSignalingStateChange?.('error')
    }
  }

  close() {
    this.intentionalClose = true
    if (this.reconnectTimer) { clearTimeout(this.reconnectTimer); this.reconnectTimer = null }
    this.stopKeepalive()
    this.cleanSessions()
    if (this.ws) { this.ws.onclose = null; this.ws.close(); this.ws = null }
  }

  private reconnect() {
    if (this.intentionalClose) return
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer)
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 16000)
    this.reconnectAttempts++
    this.reconnectTimer = setTimeout(() => {
      this.onSignalingStateChange?.('error')
      this.connect()
    }, delay)
  }

  private startKeepalive() {
    this.stopKeepalive()
    this.keepaliveTimer = setInterval(() => {
      if (!this.connected) return
      this.send('keepalive', {})
      this.keepaliveTimeout = setTimeout(() => {
        this.connected = false
        this.ws?.close()
      }, 5000)
    }, 3000)
  }

  private stopKeepalive() {
    if (this.keepaliveTimer) { clearInterval(this.keepaliveTimer); this.keepaliveTimer = null }
    if (this.keepaliveTimeout) { clearTimeout(this.keepaliveTimeout); this.keepaliveTimeout = null }
  }

  private async onMessage(msg: { type: string; data: unknown }) {
    switch (msg.type) {
      case 'peers': {
        const peers = msg.data as unknown[]
        const self = this.uid
        this.onPeerUpdate?.({ self, peers })
        break
      }
      case 'offer': {
        const data = msg.data as Record<string, unknown>
        const peerId = data['from'] as string
        const description = data['description'] as RTCSessionDescriptionInit
        const media = data['media'] as string
        const sessionId = data['session_id'] as string
        let session = this.sessions.get(sessionId)
        session = await this.createSession(session, peerId, sessionId, media)
        this.sessions.set(sessionId, session)
        await session.pc!.setRemoteDescription(new RTCSessionDescription(description))
        for (const c of session.remoteCandidates) {
          await session.pc!.addIceCandidate(new RTCIceCandidate(c))
        }
        session.remoteCandidates = []
        this.onCallStateChange?.(sessionId, 'new')
        this.onCallStateChange?.(sessionId, 'ringing')
        break
      }
      case 'answer': {
        const data = msg.data as Record<string, unknown>
        const description = data['description'] as RTCSessionDescriptionInit
        const sessionId = data['session_id'] as string
        const session = this.sessions.get(sessionId)
        await session?.pc?.setRemoteDescription(new RTCSessionDescription(description))
        this.onCallStateChange?.(sessionId, 'connected')
        break
      }
      case 'candidate': {
        const data = msg.data as Record<string, unknown>
        const candidateMap = data['candidate'] as RTCIceCandidateInit
        const sessionId = data['session_id'] as string
        let session = this.sessions.get(sessionId)
        const candidate = new RTCIceCandidate(candidateMap)
        if (session) {
          if (session.pc) {
            await session.pc.addIceCandidate(candidate)
          } else {
            session.remoteCandidates.push(candidateMap)
          }
        } else {
          session = { sid: sessionId, pid: '', pc: null, media: '', remoteCandidates: [candidateMap] }
          this.sessions.set(sessionId, session)
        }
        break
      }
      case 'bye': {
        const data = msg.data as { session_id: string }
        const sessionId = data.session_id
        this.sessions.delete(sessionId)
        this.onCallStateChange?.(sessionId, 'bye')
        break
      }
      case 'keepalive':
        if (this.keepaliveTimeout) clearTimeout(this.keepaliveTimeout)
        break
    }
  }

  private async createSession(existing: Session | undefined, peerId: string, sessionId: string, media: string): Promise<Session> {
    const session = existing || { sid: sessionId, pid: peerId, pc: null, media, remoteCandidates: [] as RTCIceCandidateInit[] }
    const turnServers = await this.getTurnCredential()
    const iceServers: RTCIceServer[] = turnServers || [{ urls: 'stun:stun.l.google.com:19302' }]

    const pc = new RTCPeerConnection({ iceServers })

    pc.onicecandidate = (e) => {
      if (e.candidate) {
        setTimeout(() => {
          this.send('candidate', {
            to: peerId,
            from: this.uid,
            candidate: e.candidate!.toJSON(),
            session_id: sessionId,
          })
        }, 1000)
      }
    }

    pc.ontrack = (e) => {
      if (e.streams[0]) {
        this._remoteStream = e.streams[0]
        this.onRemoteStream?.(e.streams[0])
      }
    }

    pc.oniceconnectionstatechange = () => {
      this.onIceConnectionState(pc.iceConnectionState)
    }

    pc.ondatachannel = () => { /* not used in Web MVP */ }

    session.pc = pc
    return session
  }

  private onIceConnectionState(state: RTCIceConnectionState) {
    const videoSender = (this._localStream?.getVideoTracks().length ?? 0) > 0
    if (!videoSender) return
    switch (state) {
      case 'connected':
      case 'completed':
        this._localStream?.getVideoTracks().forEach(t => t.enabled = true)
        break
      case 'disconnected':
      case 'failed':
        this._localStream?.getVideoTracks().forEach(t => t.enabled = false)
        break
    }
  }

  async startLocalStream() {
    try {
      const stream = await navigator.mediaDevices.getUserMedia({
        video: { width: { min: 640 }, height: { min: 480 }, frameRate: { min: 30 } },
        audio: { echoCancellation: true, noiseSuppression: true, autoGainControl: true },
      })
      this._localStream = stream
      this.onLocalStream?.(stream)
      return stream
    } catch { return null }
  }

  async startScreenShare() {
    try {
      const stream = await navigator.mediaDevices.getDisplayMedia({ video: true })
      const sender = this.pc?.getSenders().find(s => s.track?.kind === 'video')
      if (sender) await sender.replaceTrack(stream.getVideoTracks()[0])
      stream.getVideoTracks()[0].onended = () => this.stopScreenShare()
      return stream
    } catch { return null }
  }

  async stopScreenShare() {
    const stream = await this.startLocalStream()
    if (stream) {
      const sender = this.pc?.getSenders().find(s => s.track?.kind === 'video')
      if (sender) await sender.replaceTrack(stream.getVideoTracks()[0])
    }
  }

  invite(peerId: string) {
    const sessionId = `${this.uid}-${peerId}`
    this.createSession(undefined, peerId, sessionId, 'video').then(async (session) => {
      this.sessions.set(sessionId, session)
      if (this._localStream) {
        for (const track of this._localStream.getTracks()) {
          session.pc!.addTrack(track, this._localStream!)
        }
      }
      const offer = await session.pc!.createOffer()
      await session.pc!.setLocalDescription(offer)
      this.send('offer', {
        to: peerId,
        from: this.uid,
        description: { sdp: offer.sdp, type: offer.type },
        session_id: sessionId,
        media: 'video',
      })
      this.onCallStateChange?.(sessionId, 'new')
      this.onCallStateChange?.(sessionId, 'invite')
    })
  }

  async accept(sessionId: string) {
    const session = this.sessions.get(sessionId)
    if (!session || !session.pc) return
    if (this._localStream) {
      for (const track of this._localStream.getTracks()) {
        session.pc.addTrack(track, this._localStream)
      }
    }
    const answer = await session.pc.createAnswer()
    await session.pc.setLocalDescription(answer)
    this.send('answer', {
      to: session.pid,
      from: this.uid,
      description: { sdp: answer.sdp, type: answer.type },
      session_id: sessionId,
    })
    this.onCallStateChange?.(sessionId, 'connected')
  }

  reject(sessionId: string) {
    this.bye(sessionId)
  }

  bye(sessionId: string) {
    this.send('bye', { session_id: sessionId, from: this.uid })
    const session = this.sessions.get(sessionId)
    if (session?.pc) { session.pc.close(); this.sessions.delete(sessionId) }
  }

  iceRestart(sessionId: string) {
    const session = this.sessions.get(sessionId)
    if (!session?.pc) return Promise.resolve()
    return session.pc.createOffer({ iceRestart: true }).then((offer) => {
      return session.pc!.setLocalDescription(offer)
    }).then(() => {
      const desc = session.pc!.localDescription!
      this.send('offer', {
        to: session.pid,
        from: this.uid,
        description: { sdp: desc.sdp, type: desc.type },
        session_id: sessionId,
        media: session.media,
      })
    })
  }

  muteMic() {
    const audioTrack = this._localStream?.getAudioTracks()[0]
    if (audioTrack) audioTrack.enabled = !audioTrack.enabled
  }

  switchCamera() {
    // basic switch — rotate facingMode in real usage
    this.startLocalStream()
  }

  private get pc(): RTCPeerConnection | null {
    return this.sessions.values().next().value?.pc || null
  }

  private send(type: string, data: unknown) {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ type, data }))
    }
  }

  private cleanSessions() {
    if (this._localStream) {
      this._localStream.getTracks().forEach(t => t.stop())
      this._localStream = null
    }
    if (this._remoteStream) {
      this._remoteStream.getTracks().forEach(t => t.stop())
      this._remoteStream = null
    }
    this.sessions.forEach(s => s.pc?.close())
    this.sessions.clear()
  }
}
